README.md
## DHI XML
[![License: GPL v3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)
[![Build Status](https://travis-ci.org/timgabets/dhi-xml.svg?branch=master)](https://travis-ci.org/timgabets/dhi-xml)

Rust community library for serializaing/deserializing TSYS DHI (Device Host Interface) XML messages.

### Usage

```rust
use dhi_xml::{DHIRequest, DHIResponse};
```

DHI requests:
```rust
// JSON payload (for example, from HTTP request)
let iso_data = r#"{
    "i000": "0100",
    "i002": "555544******0895",
    "i007": "Transmission date & time ",
    "i011": "STAN",
    "i012": "hhmmss",
    "i013": "MMDD",
    "i037": "Retrieval Reference Number"
}"#;

// Deserializing request from the given JSON payload
let r: DHIRequest = DHIRequest::new(serde_json::from_str(&iso_data).unwrap());

// The data may now be accessed with
assert_eq!(r.iso_fields["i002"], "555544******0895");

// Serialization
let msg = r.serialize().unwrap();

// The message may be sent through the TCP stream
let s = TcpStream::connect(host);
s.write_all(&msg.as_bytes());
```

DHI response:
```rust

let s = r##"
    <?xml version="1.0"?>
    <RequestResponse>
        <Header/>
        <Result><Code>0</Code><Description>OK</Description></Result>
     	<ISO8583-87><i000>0110</i000><i002>555544******0961</i002><i003>300000</i003><i004>000000000000</i004><i007>2804114717</i007></ISO8583-87>"
    </RequestResponse>
"##;

// Deserialization from the XML payload
let resp: DHIResponse = from_reader(s.as_bytes()).unwrap();

// Accessing data
assert_eq!(resp.res.code, 0);
assert_eq!(resp.res.description, "OK");
assert_eq!(resp.iso_fields["i000"], "0110");
assert_eq!(resp.iso_fields["i002"], "555544******0961");

// Serializing response to JSON
let msg = resp.serialize().unwrap();

// Sending as payload in HTTP response
Ok(HttpResponse::Ok()
    .content_type("application/json")
    .header("X-Hdr", "sample")
    .body(msg)),
```



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.1.3] - 2020-06-04
### Changed
- Using MIT license instead of GPL.

## [0.1.2] - 2020-06-02
### Added
- README use cases.

## [0.1.1] - 2020-06-01
### Added
- The library code moved from the original [lakgves repo](https://github.com/timgabets/lakgves).

## [0.1.0] - 2020-06-01
### Added
- Initial commit.


src/lib.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;
use serde_xml_rs::to_string;

use std::collections::BTreeMap;

mod util;

#[derive(Serialize, Debug)]
pub struct Header {
    #[serde(rename(serialize = "MessageID"))]
    message_id: i64,
    #[serde(rename(serialize = "SystemID"))]
    system_id: String,
}

#[derive(Serialize, Debug)]
#[serde(rename(serialize = "RequestInput"))]
pub struct DHIRequest {
    #[serde(rename(serialize = "ISO8583-87"))]
    iso_fields: Value,
}

impl DHIRequest {
    pub fn new(iso_obj: Value) -> Self {
        let mut req = DHIRequest {
            iso_fields: iso_obj,
        };
        if req.iso_fields.get("i007") == None {
            req.iso_fields["i007"] = serde_json::value::Value::String(util::get_mmddhhmmss());
        }
        if req.iso_fields.get("i011") == None {
            req.iso_fields["i011"] = serde_json::value::Value::String(util::gen_stan());
        }
        if req.iso_fields.get("i012") == None {
            req.iso_fields["i012"] = serde_json::value::Value::String(util::get_hhmmss());
        }
        if req.iso_fields.get("i013") == None {
            req.iso_fields["i013"] = serde_json::value::Value::String(util::get_mmdd());
        }
        if req.iso_fields.get("i037") == None {
            req.iso_fields["i037"] = serde_json::value::Value::String(util::gen_rrn());
        }
        req
    }

    pub fn serialize(&self) -> Result<String, serde_xml_rs::Error> {
        let serialized = to_string(&self)?;
        let serialized = format!("{:05}{}", serialized.len(), serialized);
        Ok(serialized)
    }
}

#[derive(Deserialize, Debug)]
#[serde(rename(deserialize = "Result"))]
#[serde(rename_all = "PascalCase")]
pub struct DHIResult {
    code: i32,
    description: String,
}

#[derive(Deserialize, Debug)]
#[serde(rename(deserialize = "RequestResponse"))]
pub struct DHIResponse {
    #[serde(rename(deserialize = "Result"))]
    res: DHIResult,

    #[serde(rename(deserialize = "ISO8583-87"))]
    iso_fields: BTreeMap<String, String>,
}

impl DHIResponse {
    pub fn serialize(&self) -> Result<String, serde_json::error::Error> {
        let serialized = serde_json::to_string(&self.iso_fields)?;
        Ok(serialized)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_xml_rs::from_reader;

    #[test]
    fn dhi_request_new_existing_fields() {
        let iso_data = r#"{
            "i000": "0100",
            "i002": "555544******0895",
            "i007": "Transmission date & time ",
            "i011": "STAN",
            "i012": "hhmmss",
            "i013": "MMDD",
            "i037": "Retrieval Reference Number"
        }"#;

        let r: DHIRequest = DHIRequest::new(serde_json::from_str(&iso_data).unwrap());
        assert_eq!(
            r.iso_fields["i007"], "Transmission date & time ",
            "i007 original value should be left untouched"
        );
        assert_eq!(
            r.iso_fields["i011"], "STAN",
            "i011 original value should be left untouched"
        );
        assert_eq!(
            r.iso_fields["i012"], "hhmmss",
            "i012 original value should be left untouched"
        );
        assert_eq!(
            r.iso_fields["i013"], "MMDD",
            "i013 original value should be left untouched"
        );
        assert_eq!(
            r.iso_fields["i037"], "Retrieval Reference Number",
            "Provided RRN should be left untouched"
        );
    }

    #[test]
    fn dhi_response_deserialization() {
        let s = r##"
        <?xml version="1.0"?>
        <RequestResponse>
            <Header/>
            <Result><Code>0</Code><Description>OK</Description></Result>
            <ISO8583-87><i000>0110</i000><i002>555544******0961</i002><i003>300000</i003><i004>000000000000</i004><i007>2804114717</i007><i043><![CDATA[IDDQD AM. 341215574     341215574 MSKRU]]></i043><i120>UD038IR0044444CR009ES0048100IA0103510198686</i120></ISO8583-87>"
        </RequestResponse>
        "##;

        let resp: DHIResponse = from_reader(s.as_bytes()).unwrap();

        assert_eq!(resp.res.code, 0);
        assert_eq!(resp.res.description, "OK");

        assert_eq!(resp.iso_fields["i000"], "0110");
        assert_eq!(resp.iso_fields["i002"], "555544******0961");
        assert_eq!(resp.iso_fields["i003"], "300000");
        assert_eq!(resp.iso_fields["i004"], "000000000000");
        assert_eq!(resp.iso_fields["i007"], "2804114717");
        assert_eq!(
            resp.iso_fields["i043"],
            "IDDQD AM. 341215574     341215574 MSKRU"
        );
        assert_eq!(
            resp.iso_fields["i120"],
            "UD038IR0044444CR009ES0048100IA0103510198686"
        );

        assert_eq!(resp.serialize().unwrap(), "{\"i000\":\"0110\",\"i002\":\"555544******0961\",\"i003\":\"300000\",\"i004\":\"000000000000\",\"i007\":\"2804114717\",\"i043\":\"IDDQD AM. 341215574     341215574 MSKRU\",\"i120\":\"UD038IR0044444CR009ES0048100IA0103510198686\"}");
    }
}
